src/MonoMod.Core/Platforms/PlatformTriple.cs

```diff
@@ -626,106 +626,234 @@ public MethodBase GetRealDetourTarget(MethodBase from, MethodBase to)
-
-                // only if the return class is ByRef do we need to do something
-                // TODO: perform better decisions based on the ABI argument order and return class
-                if (returnClass == TypeClassification.ByReference)
+                switch (argumentKinds[i])
                 {
-                    var thisType = from.GetThisParamType();
-                    var retPtrType = retType.MakeByRefType();
-
-                    var newRetType = Abi.ReturnsReturnBuffer ? retPtrType : typeof(void);
-
-                    int thisPos = -1, retBufPos = -1, argOffset = -1;
-
-                    var paramList = from.GetParameters();
+                    case SpecialArgumentKind.ThisPointer when hasThis:
+                        thisPos = argumentTypes.Count;
+                        argumentTypes.Add(from.GetThisParamType());
+                        break;
+
+                    case SpecialArgumentKind.ReturnBuffer when hasReturnBuffer:
+                        returnBufferPos = argumentTypes.Count;
+                        argumentTypes.Add(returnBufferType);
+                        break;
+
+                    case SpecialArgumentKind.GenericContext when requiresGenericContextFixup:
+                        // Currently, we do the bare minimum: we acknowledge that
+                        // the generic context exists. That's all. After that,
+                        // we simply throw it out of the window and rely on
+                        // the generic context (if any) baked into in the callee.
+                        //
+                        // While this does work fine, it introduces funny little
+                        // holes in the type-safety premise of .NET:
+                        //
+                        // // This may return `false`!
+                        // bool Hook<T>(T it)
+                        //     => typeof(T).IsAssignableFrom(it.GetType());
+                        //
+                        // This behavior is caused by the fact that constructed
+                        // generics for reference types are Javad into a single
+                        // definition at runtime (and rightfully so).
+                        // So, even if you try to detour a specific generic
+                        // implementation using something like
+                        // `to.MakeGenericMethod(typeof(string))`,
+                        // your hook will receive calls for all
+                        // reference type-based implementations out there.
+                        // However, since we do not patch the generic context
+                        // of the provided hook, it remains unchanged,
+                        // causing `typeof(T)` to defy users' expectations.
+                        //
+                        // So, TODO: patch the generic context of the detour target
+                        // and/or introduce a call filter based on the current context.
```

Comment:
This change in behavior (towards silently not doing what the user wants) drives me to want to also block generics in the RuntimeDetour API (though I'm OK with leaving it open from the Core API).

Reply:
That's probably where I'd put it, yes.